File: backend/app/models/emi.py
import uuid
from datetime import datetime, date
from sqlalchemy import Column, String, DateTime, Date, Integer, Numeric, Enum as SQLEnum, ForeignKey, Text
from app.core.types import UUID
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class ContractStatus(str, enum.Enum):
    ACTIVE = "active"
    COMPLETED = "completed"
    DEFAULTED = "defaulted"
    CANCELLED = "cancelled"


class PaymentStatus(str, enum.Enum):
    PENDING = "pending"
    PAID = "paid"
    PARTIAL = "partial"
    OVERDUE = "overdue"
    WAIVED = "waived"


class EMIContract(Base):
    __tablename__ = "emi_contracts"
    
    id = Column(UUID(), primary_key=True, default=uuid.uuid4)
    contract_number = Column(String(50), unique=True, nullable=False, index=True)
    
    # Customer and Device
    customer_id = Column(UUID(), ForeignKey("customers.id"), nullable=False)
    device_id = Column(UUID(), ForeignKey("devices.id"), nullable=False)
    
    # Product Details
    product_name = Column(String(255))
    product_price = Column(Numeric(12, 2), nullable=False)
    down_payment = Column(Numeric(12, 2), default=0)
    
    # EMI Details
    principal_amount = Column(Numeric(12, 2), nullable=False)
    interest_rate = Column(Numeric(5, 2), default=0)  # Annual interest rate
    tenure_months = Column(Integer, nullable=False)
    emi_amount = Column(Numeric(10, 2), nullable=False)
    total_amount = Column(Numeric(12, 2), nullable=False)
    
    # Dates
    start_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=False)
    emi_due_day = Column(Integer, default=1)  # Day of month when EMI is due
    
    # Status
    status = Column(SQLEnum(ContractStatus), default=ContractStatus.ACTIVE, index=True)
    grace_period_days = Column(Integer, default=7)
    
    # Paid Amount Tracking
    total_paid = Column(Numeric(12, 2), default=0)
    emis_paid = Column(Integer, default=0)
    
    # Notes
    notes = Column(Text)
    
    # Relations
    customer = relationship("Customer", back_populates="contracts")
    device = relationship("Device")
    payments = relationship("EMIPayment", back_populates="contract")
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(UUID(), ForeignKey("users.id"))
    
    def __repr__(self):
        return f"<EMIContract {self.contract_number}>"


class EMIPayment(Base):
    __tablename__ = "emi_payments"
    
    id = Column(UUID(), primary_key=True, default=uuid.uuid4)
    
    contract_id = Column(UUID(), ForeignKey("emi_contracts.id"), nullable=False)
    installment_number = Column(Integer, nullable=False)
    
    # Amount
    due_amount = Column(Numeric(10, 2), nullable=False)
    paid_amount = Column(Numeric(10, 2), default=0)
    late_fee = Column(Numeric(10, 2), default=0)
    
    # Dates
    due_date = Column(Date, nullable=False, index=True)
    paid_date = Column(DateTime)
    
    # Status
    status = Column(SQLEnum(PaymentStatus), default=PaymentStatus.PENDING, index=True)
    
    # Payment Reference
    payment_method = Column(String(50))  # cash, upi, bank_transfer, etc.
    payment_reference = Column(String(100))
    
    # Relations
    contract = relationship("EMIContract", back_populates="payments")
    
    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    recorded_by = Column(UUID(), ForeignKey("users.id"))
    
    def __repr__(self):
        return f"<EMIPayment {self.contract_id} - {self.installment_number}>"

File: backend-node/src/models/emi.js
import { randomUUID } from 'crypto';
import { DataTypes } from 'sequelize';
import { sequelize } from '../db.js';
import { enumField, ContractStatus, PaymentStatus } from '../utils/enums.js';

export const EMIContract = sequelize.define(
  'EMIContract',
  {
    id: { type: DataTypes.STRING(36), primaryKey: true, defaultValue: randomUUID },
    contract_number: { type: DataTypes.STRING(50), allowNull: false, unique: true },
    customer_id: { type: DataTypes.STRING(36), allowNull: false },
    device_id: { type: DataTypes.STRING(36), allowNull: false },
    product_name: { type: DataTypes.STRING(255) },
    product_price: { type: DataTypes.DECIMAL(12, 2), allowNull: false },
    down_payment: { type: DataTypes.DECIMAL(12, 2), defaultValue: 0 },
    principal_amount: { type: DataTypes.DECIMAL(12, 2), allowNull: false },
    interest_rate: { type: DataTypes.DECIMAL(5, 2), defaultValue: 0 },
    tenure_months: { type: DataTypes.INTEGER, allowNull: false },
    emi_amount: { type: DataTypes.DECIMAL(10, 2), allowNull: false },
    total_amount: { type: DataTypes.DECIMAL(12, 2), allowNull: false },
    start_date: { type: DataTypes.DATEONLY, allowNull: false },
    end_date: { type: DataTypes.DATEONLY, allowNull: false },
    emi_due_day: { type: DataTypes.INTEGER, defaultValue: 1 },
    status: enumField({ field: 'status', defaultValue: 'active' }),
    grace_period_days: { type: DataTypes.INTEGER, defaultValue: 7 },
    total_paid: { type: DataTypes.DECIMAL(12, 2), defaultValue: 0 },
    emis_paid: { type: DataTypes.INTEGER, defaultValue: 0 },
    notes: { type: DataTypes.TEXT },
    created_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    updated_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    created_by: { type: DataTypes.STRING(36) },
  },
  { tableName: 'emi_contracts', timestamps: false },
);

export const EMIPayment = sequelize.define(
  'EMIPayment',
  {
    id: { type: DataTypes.STRING(36), primaryKey: true, defaultValue: randomUUID },
    contract_id: { type: DataTypes.STRING(36), allowNull: false },
    installment_number: { type: DataTypes.INTEGER, allowNull: false },
    due_amount: { type: DataTypes.DECIMAL(10, 2), allowNull: false },
    paid_amount: { type: DataTypes.DECIMAL(10, 2), defaultValue: 0 },
    late_fee: { type: DataTypes.DECIMAL(10, 2), defaultValue: 0 },
    due_date: { type: DataTypes.DATEONLY, allowNull: false },
    paid_date: { type: DataTypes.DATE },
    status: enumField({ field: 'status', defaultValue: 'pending' }),
    payment_method: { type: DataTypes.STRING(50) },
    payment_reference: { type: DataTypes.STRING(100) },
    created_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    updated_at: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    recorded_by: { type: DataTypes.STRING(36) },
  },
  { tableName: 'emi_payments', timestamps: false },
);

export { ContractStatus, PaymentStatus };
export default EMIContract;
